File: app/api/schemas/compare.py
"""Pydantic schemas for /api/v1/compare endpoint."""

from pydantic import BaseModel, Field


class CompareRequest(BaseModel):
    prompt: str = Field(..., min_length=10, description="Prompt / task for all models")
    system_prompt: str = Field(default="", description="Optional system prompt")
    model_ids: list[str] | None = Field(
        default=None,
        description="Subset of model IDs to run. None = all models.",
    )
    free_only: bool = Field(default=False, description="Run only free-tier models")


class EvalBreakdown(BaseModel):
    length: int
    first_person: int
    digits: int
    structure: int
    no_opener: int
    readability: int
    ai_isms_penalty: int


class ModelResultResponse(BaseModel):
    model_id: str
    display_name: str
    provider: str
    is_free: bool
    text: str
    elapsed_sec: float
    error: str
    score: int | None
    zone: str | None        # "Green" | "Orange" | "Red" | None (on error)
    breakdown: EvalBreakdown | None
    ai_isms_found: list[str]


class CompareResponse(BaseModel):
    results: list[ModelResultResponse]
    total_models: int
    successful: int
    prompt: str

File: frontend/src/api/compare.ts
import api from './client'

export interface CompareRequest {
  prompt: string
  system_prompt?: string
  model_ids?: string[] | null
  free_only?: boolean
}

export interface EvalBreakdown {
  length: number
  first_person: number
  digits: number
  structure: number
  no_opener: number
  readability: number
  ai_isms_penalty: number
}

export interface ModelResult {
  model_id: string
  display_name: string
  provider: string
  is_free: boolean
  text: string
  elapsed_sec: number
  error: string
  score: number | null
  zone: 'Green' | 'Orange' | 'Red' | null
  breakdown: EvalBreakdown | null
  ai_isms_found: string[]
}

export interface CompareResponse {
  results: ModelResult[]
  total_models: number
  successful: number
  prompt: string
}

export async function runComparison(req: CompareRequest): Promise<CompareResponse> {
  // Explicit path relative to baseURL (/api), no leading slash to avoid double /api
  const { data } = await api.post<CompareResponse>('v1/compare', req)
  return data
}
